orocos_kdl/src/chainhdsolver_vereshchagin.cpp

```diff
@@ -0,0 +1,522 @@
+{
+    unsigned int j = 0;
+
+    for (unsigned int i = 1; i <= ns; i++)
+    {
+        segment_info& s = results[i];
+        //Calculation of joint and segment accelerations
+        //equation g) qdotdot[i] = D^-1*(Q - Z'(R + P(C + acc[i-1]) + E*nu))
+        // = D^-1(u - Z'(P*acc[i-1] + E*nu)
+        Twist a_g;
+        Twist a_p;
+        if (i == 1)
+        {
+            a_p = acc_root;
+        }
+        else
+        {
+            a_p = results[i - 1].acc;
+        }
+
+        //The contribution of the constraint forces at segment i
+        Vector6d tmp = s.E*nu;
+        Wrench constraint_force = Wrench(Vector(tmp(3), tmp(4), tmp(5)),
+                                         Vector(tmp(0), tmp(1), tmp(2)));
+
+        //acceleration components are also computed
+        //Contribution of the acceleration of the parent (i-1)
+        Wrench parent_force = s.P*a_p;
+        double parent_forceProjection = -dot(s.Z, parent_force);
+        double parentAccComp = parent_forceProjection / s.D;
+
+        //The constraint force and acceleration force projected on the joint axes -> axis torque/force
+        constraint_torques(j) = -dot(s.Z, constraint_force);
+        //The result should be the torque originating from the end-effector constraints
+
+        // Total torque on the joint resulting from the parent forces, constraint forces and nullspace forces.
+        total_torques(j) = s.u + parent_forceProjection + constraint_torques(j);
+        // q_dotdot(j) is also equal to: total_torques(j) / s.D
+
+        s.constAccComp = constraint_torques(j) / s.D;
+        s.nullspaceAccComp = s.u / s.D;
+
+        // Total joint space acceleration resulting from accelerations of parent joints, constraint forces and nullspace forces.
+        q_dotdot(j) = (s.nullspaceAccComp + parentAccComp + s.constAccComp);
+        // Returns segment's spatial acceleration in link distal-tip coordinates. For use needs to be transformed
+        s.acc = s.F.Inverse(a_p + s.Z * q_dotdot(j) + s.C);
+        if (chain.getSegment(i - 1).getJoint().getType() != Joint::Fixed)
+            j++;
+    }
+}
+
+// Returns Cartesian acceleration of links in robot base coordinates
+void ChainHdSolver_Vereshchagin::getTransformedLinkAcceleration(Twists& x_dotdot)
+{
+    assert(x_dotdot.size() == ns + 1);
+    x_dotdot[0] = acc_root;
+    for (unsigned int i = 1; i < ns + 1; i++)
+        x_dotdot[i] = results[i].F_base.M * results[i].acc;
+}
+
+// Returns total torque acting on each joint (constraints + nature + external forces)
+void ChainHdSolver_Vereshchagin::getTotalTorque(JntArray &total_tau)
+{
+    assert(total_tau.data.size() == total_torques.size());
+    total_tau.data = total_torques;
+}
+
+// Returns magnitude of the constraint forces acting on the end-effector: Lagrange Multiplier
+void ChainHdSolver_Vereshchagin::getContraintForceMagnitude(Eigen::VectorXd &nu_)
+{
+    assert(nu_.size() == nu.size());
+    nu_ = nu;
+}
+
+/*
```

Comment:
Any reason why these commented functions are still here?

Reply:
As I mentioned in the above description of the pull-request and in https://github.com/DjoleMNE/orocos_kinematics_dynamics/blob/29087bde5ee092170e552176a6a2c599e746f37c/orocos_kdl/tests/solvertest.cpp#L951, these getters are added for the solver's final output and this information is useful for the state- simulation and estimation purposes. This algorithm provides useful features for not-only control use-cases but also for the above-mentioned one.

Reply:
If you refer to the functions that are located below the "getContraintForceMagnitude()" method, then for those functions, I think there were used by Azamat Shakhimardanov while performing experiments for his PhD work. Nevertheless, I think the reason why hi and/or @smits left them there is the fact that these functions could be useful for the users in certain robotics use cases since the algorithm provides useful features for control, simulation and state-estimation purposes. I think that the first 9 functions could indeed be frequently used by KDL users, however, for the rest of the functions I cannot say the same, since those are very specific.

Reply:
Nevertheless, I would suggest keeping those functions there. However, I did not test those functions, except for "getLinkCartesianAcceleration()" method which is similar to the "getTransformedLinkAcceleration()" method that I added. 

For them, i) it would be nice to have the appropriate tests in the solvertest.cpp file and ii) do some renaming, since some of the current function names are ambiguous (cf. getLinkCartesianVelocity() vs. getLinkVelocity() method names).

I would suggest a separate pull request for these particular improvements.